### PR TITLE
Switch the DRBGs from AES128 to AES256

### PR DESCRIPTION
**Pull request by kroeckx**

### CODE REVIEW

File: include/internal/rand.h [hunk 1]
```diff
@@ -15,14 +15,22 @@
 
 /*
  * Default security strength (in the sense of [NIST SP 800-90Ar1])
- * of the default OpenSSL DRBG, and the corresponding NID.
  *
- * Currently supported values: 128, 192, 256
+ * NIST SP 800-90Ar1 supports the strength of the DRBG being smaller than that
+ * of the cipher by collecting less entropy. The current DRBG implemantion does
+ * not take RAND_DRBG_STRENGTH into account and sets the strength of the DRBG
+ * to that of the cipher.
```
Comment (mspncp) [inline include/internal/rand.h]: This is not true. `drbg->strength` is passed as third argument to to the `get_entropy` callback which implies that a minimum amount of `RAND_DRBG_STRENGTH` entropy bits is requested for (re-)seeding.

File: include/internal/rand.h [hunk 2]
```diff
@@ -15,14 +15,22 @@
...
  *
- * TODO(DRBG): would be nice to have the strength configurable
+ * RAND_DRBG_STRENGTH is currently only used for the legacy RAND
+ * implementation.
+ *
+ * Currently supported ciphers are: NID_aes_128_ctr, NID_aes_192_ctr and
+ * NID_aes_256_ctr
+ *
```
Comment (mspncp) [inline include/internal/rand.h]: These comments should be reverted. The original statement was about security strength, not about supported ciphers.

**Comment by paulidale** [comment]
Does this change the strength?  NIST SP 800-90Ar1 claims in section 8.4 that:

> _Security_strength_of_output_ = **min**(_output_length_, _DRBG_security_strength_)

AES128 and ASE256 both have a 128 bit output length which limits the strength that can be claimed to 128 bits.

With the lack of any practical attack against either flavour of AES, I'd usually pick the faster algorithm.
Alternatively, allow the user to choose any 128 bit block cipher.

**Comment by mspncp** [comment]
> AES128 and AES256 both have a 128 bit output length which limits the strength that can be claimed to 128 bits.

The security strength is not limited by the block length, it depends on the size of the internal key. If the former would be the case then having AES256 would not make sense in the first place. In 'Table 3: Definitions for the CTR_DRBG' on page 49 the security strengths for the CTR_DRBG are explicitely listed. So when you request 256 bits of random output, you will get two AES blocks, but the security strength of the output will be 128 for AES-CTR 128 and 256 for AES-CTR 256.

**Comment by mspncp** [comment]
Apropos AES256: I would prefer if the title of the commit log would be more precise about that and say

> Switch the DRBGs from AES-128-CTR to AES-256-CTR

following our conventions:

```
crypto/objects/obj_dat.h:1943:    {"AES-128-CTR", "aes-128-ctr", NID_aes_128_ctr},
crypto/objects/obj_dat.h:1944:    {"AES-192-CTR", "aes-192-ctr", NID_aes_192_ctr},
crypto/objects/obj_dat.h:1945:    {"AES-256-CTR", "aes-256-ctr", NID_aes_256_ctr},
```

**Comment by paulidale** [comment]
I suspect this is another question during validation....  I know of a (validated) toolkit that does limit the strength to the block size.

File: include/internal/rand.h (hunk 1 above)
Comment (kroeckx) [reply include/internal/rand.h]: But drbg->strength is not based on RAND_DRBG_STRENGTH but on RAND_DRBG_NID, it's keylen * 8. This is not required by NIST SP 800-90. With AES-256-CTR you should be able to set it to 128, 192 or 256, but the code currently doesn't allow this.

Comment (mspncp) [reply include/internal/rand.h]: Oh sorry, you're right. It's ok if you leave the comment here. But please fix the typo: implemantion -> implementation.

**Comment by mspncp** [comment]
And it would be nice if you would change the subject line as suggested to

> Switch the DRBGs from AES-128-CTR to AES-256-CTR

**Pull request closed without merging**